### ISSUE TITLE
Std::array square bracket assignment not working

### ISSUE THREAD

**Issue by neutralid** [pybind/pybind11#1134]
Possibly the root of my issue is the lack of apparent documentation on array/pointer binding. If there is such a thing, I would be grateful to be pointed to it.

We have an older code base (and older developers) and have a lot of code using fixed buffers and pointers for dynamic memory.

It's hard to provide a complete working example here, but essentially we have

class SomeClass; // Doesn't much matter what is in this class, I don't think

class AnotherClass {
SomeClass* arrayOfPointers[100];
}

and in python
x=SomeClass()
y=AnotherClass()
y.arrayOfPointers[0]=x

In short, I wanted to instantiate SomeClass in python, then set one of the pointers to that instance so that it is accessible to the class.

The only mention of arrays I could find mentioned using std::array was supported, so I converted to that to avoid the headache of setter/getters (though, with a lot of old code to support it would be nice to know if there was a way to expose a regular pointer/array to python and make it compatible with square bracket notation/__setitem__/__getitem__ but per member variable, not for the class).

Anyways, that seemed to compile and run and y.arrayOfPointers[0] returns "None" as expected. Trying to assign it (y.arrayOfPointers[0]=x), however, does not seem to work. No error messages, but when I try to read after writing, it continues to show None. I could create a new list, set list[0] to x, and then assign the new list to arrayOfPointers as a whole, though, and then it seemed to report back correctly. 

z=[None] * 100
z[0] = x
y.arrayOfPointers = z

In short:
1) Is there something special I need to do to get std::array working correctly for single-member assignment with square brackets, or is this a true bug?

2) Is there a way to expose a c-style fixed array (e.g., not a std::array<>) without ugly code/getters/setters that is compatible with square brackets?

3) Same question as 2, but for a regular pointer used for an old-school dynamically allocated array that would need bounds populated by code?

4) Can someone throw a section about array binding into the docs? Seems like a very common use case and the breadcrumbs are hard to follow. I could only find a couple examples in one of the test scripts which used std::array. I see a lot of mention of NumPy and I'm not sure if that applies for what I need, generic lists of objects/object pointers, I thought that was particularly for number-crunching heavy programs.

Thanks in advance!

**Comment by neutralid** [comment]
Thank you so much for the information. I had missed the nuance in the documentation that STL containers are only assignable in their entirety unless you explicitly bind them as opaque objects.

For 2 and 3, this seems like an unfortunate limitation. Previously, my code featured SIP bindings, which allowed you to specify arbitrary "get" and "set" code for bound objects, so you could write accessors for primitive pointers/arrays that turned them into standard list objects without pulling in numpy. I may explore writing a simple wrapper class for primitives with a bound __setitem__ and __getitem__ method that hopefully will be a little more intuitive (at least, for me).

For 4, I was referring to the lack of documentation on binding c-style/non-STL arrays or pointers-as-arrays. Following the docs explanations for standard class and function bindings, it seemed like a logical step to me to be able to bind "int array[100]" by writing .def("array", &MyClass::array). 

For dynamically allocated memory, I can intuitively tell there's no good way to get a size to create a list, but it still doesn't leave me with a clear picture of what I have to do to bind it. From what I could find, others simply wrote full set/get methods for the arrays in the c++ code and bound those instead, so I think I'm not the only one facing this question. It would be great to have my hand held and told "this is exactly what you do to bind int[100] to a Python list". You suggest numpy, which is perhaps the "correct" solution, but for someone newly coming into pybind11, I assumed numpy was for heavier duty number manipulation and didn't explore it as a solution for my problem.

**Comment by YannickJadoul** [comment]
I see, `py::bind_vector` only works for vector-like classes. But the thing is, a C array is not a full Python list equivalent. For example: what would you do if someone would call for example `y.arrayOfPointers.append(...)`?

**Comment by neutralid** [comment]
We just accepted that it wouldn't yield working code on the cpp side, and since our getters/setters had bounds checking, if you actually tried to access one of those elements it would throw.
